include/exadg/operators/variable_coefficients.h

```diff
@@ -23,168 +23,231 @@
 
-    coefficients_face.reinit(matrix_free.n_inner_face_batches() +
-                               matrix_free.n_boundary_face_batches(),
-                             matrix_free.get_n_q_points_face(quad_index));
+    if(use_face_loops)
+      coefficients_face.reinit(matrix_free.n_inner_face_batches() +
+                                 matrix_free.n_boundary_face_batches(),
+                               matrix_free.get_n_q_points_face(quad_index));
+
+    if(use_neighbor_faces)
+      coefficients_face_neighbor.reinit(matrix_free.n_inner_face_batches(),
+                                        matrix_free.get_n_q_points_face(quad_index));
 
-    coefficients_face_neighbor.reinit(matrix_free.n_inner_face_batches(),
-                                      matrix_free.get_n_q_points_face(quad_index));
+    if(use_cell_based_face_loops)
+    {
+      unsigned int const n_faces_per_cell =
+        matrix_free.get_dof_handler().get_triangulation().get_reference_cells()[0].n_faces();
 
-    // TODO
-    // // cell-based face loops
-    // coefficients_face_cell_based.reinit(matrix_free.n_cell_batches()*2*dim,
-    // matrix_free.get_n_q_points_face(quad_index));
+      coefficients_face_cell_based.reinit(matrix_free.n_cell_batches() * n_faces_per_cell,
+                                          matrix_free.get_n_q_points_face(quad_index));
+    }
   }
 
+  /**
+   * Fills the tables with a constant coefficient.
+   */
   void
   fill(coefficient_type const & constant_coefficient)
   {
     coefficients_cell.fill(constant_coefficient);
-    coefficients_face.fill(constant_coefficient);
-    coefficients_face_neighbor.fill(constant_coefficient);
 
-    // TODO
-    // coefficients_face_cell_based.fill(constant_coefficient);
-  }
+    if(use_face_loops)
+      coefficients_face.fill(constant_coefficient);
 
-  // variable coefficients
+    if(use_neighbor_faces)
+      coefficients_face_neighbor.fill(constant_coefficient);
 
-  // cell
+    if(use_cell_based_face_loops)
+      coefficients_face_cell_based.fill(constant_coefficient);
+  }
+
+  //! Coefficient table for cells
   dealii::Table<2, coefficient_type> coefficients_cell;
 
-  // face-based loops
+  //! Coefficient table for faces
   dealii::Table<2, coefficient_type> coefficients_face;
+
+  //! Coefficient table for neighbor faces
   dealii::Table<2, coefficient_type> coefficients_face_neighbor;
 
-  // TODO
-  //  // cell-based face loops
-  //  dealii::Table<2, coefficient_type> coefficients_face_cell_based;
+  //! Coefficient table for faces with cell-based access
+  dealii::Table<2, coefficient_type> coefficients_face_cell_based;
+
+  //! Boolean switch to use a coefficient table for faces
+  bool use_face_loops{false};
+
+  /**
+   * @brief Boolean switch to use a coefficient table for neighbors, in case the coefficients
+   * can differ on the two sides of a face.
+   */
+  bool use_neighbor_faces{false};
```

Comment:
I can not see a use case in the current code where face loops are used, but no neighboring data. So I wonder whether we should introduce this flexibility now. Similarly, it looks like the current code allows the use of neighboring faces, but `use_face_loops` being deactivated. From a logical perspective, I believe this boolean should be nested into the other boolean `use_face_loops`.

Reply:
One use case is if the coefficients are computed from an at least C0-continuous function of just space and time, like the field functions. In that case the coefficients are the same on both sides and it wouldn't be necessary to store them twice. Isn't that right?

Reply:
Cell-based face loops always use neighboring data, right? This dependency should also be reflected in the implementation. I am thinking of the following logic:

```c++
if(face_loops)
{
  if(use_cell_based_face_loops)
  {
    
  }
  else // when using a separate loop over faces in MatrixFree implementation
  {
    ...

    if(use_neighbor_faces)
    {
    }
  }
}
```
As said above, I would not implement `use_neighbor_faces` for now because of missing use cases. The implementation of a spatial PDE operator decides very naturally whether to read neighboring data or not. Hence, in terms of data transfer from main memory (as a main factor for computational efficiency) it should not be relevant to introduce the optimization `use_neighbor_faces=false`. One could see a benefit in storage requirements, but this is typically less of a concern in our applications with dealii's matrix-free implementation. Or do I miss something?

Reply:
I'd be fine with removing the optimization here and in the operators just have:

```
auto const coefficient = get_coefficient_face(face, q);
```

and omit the additional call to `get_coefficient_face_neighbor(face,q)` if it is not needed, i.e. if it is a priori known that the coefficients are continuous. Same is true calculating and setting the coefficients at every time step for every discretization.

This is what you mean, right? Because then exactly we would primarily benefit from storage.

Reply:
Yes exactly.